Adding a Getting Started page (#214)
haskell-infra/www.haskell.org: site/get-started.markdown

---
title: Get Started
page: get-started
isGettingStarted: true
---

# Get started
Welcome, new Haskeller! Read on to quickly set up your Haskell dev environment, execute your first lines of code, and get directions for further learning!

## Content
  - [Set up Haskell dev environment](#set-up-haskell-dev-environment)
    - [GHCup: universal installer](#ghcup-universal-installer)
    - [Editor](#editor)
  - [Running first lines of code](#running-first-lines-of-code)
  - [Writing your first Haskell program](#writing-your-first-haskell-program)
  - [Join the community](#join-the-community)
  - [Next steps](#next-steps)

> ## Set up Haskell dev environment

Reviewer: Under this header I'd add a little text like (with better phrasing as mine):

```
A complete haskell dev environment consist in the haskell tooling (compiler, language server client and build tool) and an editor compatible with the language server protocol. The quickest way to get all of those is:

- Use Ghcup to manage all haskell tooling
- Use vscode as your editor with the haskell extension.
```

I think it is important to anticipate how many things are involved, otherwise it isn't clear how much of the guide should I read. 

Let me know what you think

Author: Sure, sounds good, added it!

### GHCup: universal installer

[GHCup](https://www.haskell.org/ghcup/#) is a universal installer for Haskell that will install for you everything you need to program in Haskell, and then will also help you manage those installations in the future (update, switch versions, ...).

Follow instructions at [GHCup webpage](https://www.haskell.org/ghcup/#) to install GHCup. Then, use it to install the Haskell Toolchain, which consists of:

1. **GHC** -> Haskell compiler. We will use it below to run our examples, but in practice, you will mostly be using a build tool like `cabal` or `Stack` to build your code, instead of `GHC` directly.
2. **HLS** -> Haskell Language Server -> You won't use this directly, instead your code editor will use it in the background to provide you with a great experience while editing Haskell code.
3. **cabal** -> Haskell build tool -> You will use this to structure your Haskell projects, build them, run them, define dependencies, ...

> 4. **Stack** -> Haskell build tool -> alternative to `cabal`

Reviewer: Notice that `stack` needs some special configuration to integrate with `hls`. Actually, `ghcup` warn you about this when installing `stack`. If newcommer skip the configuration step, `hls` will complain about "could not fin ghc" or something like that. I've already seen a few reddit/SO question about this, so maybe it worths to point this out. 

The ghcup output looks like this:
```
[ Info  ] verifying digest of: stack-2.9.1-linux-x86_64-static.tar.gz
[ Info  ] Unpacking: stack-2.9.1-linux-x86_64-static.tar.gz to /home/<username>/.ghcup/tmp/ghcup-e994793785bef45f
[ Info  ] Installing stack
[ Info  ] Stack manages GHC versions internally by default. In order to make it use ghcup installed
[ ...   ] GHC versions you can run the following commands:
[ ...   ]   stack config set install-ghc false --global
[ ...   ]   stack config set system-ghc  true  --global
[ ...   ] 
[ ...   ] On windows, you may find the following config options useful too:
[ ...   ]   skip-msys, extra-path, extra-include-dirs, extra-lib-dirs
[ ...   ] 
[ ...   ] Also check out: https://docs.haskellstack.org/en/stable/yaml_configuration
[ ...   ] 
[ ...   ] !!! Additionally, you should upgrade stack only through ghcup and not use 'stack upgrade' !!!
[ ...   ] 
Success
```

Also, new `stack` version allows for instalation hooks, but I don't know it `ghcup` uses that by default yet (I think it is comming) @hasufell might help here.

Author: I could use some help with figuring out what we write about this. @hasufell , any immediate opinion on this?

Of other editors that have good Haskell extensions, the most popular ones are Vim and Emacs.

## Running first lines of code

We have everything set up, let's use it!

`GHC` brings an interactive interpreter called `GHCi` together with it, which is great for playing with Haskell and trying things out, so let's give it a spin.

Run `ghci`, which should start a new prompt for you.

Let's do a simple calculation to check Haskell's computing capabilities:
```
> 6 + 3^2 * 4
42
```

42 is a nice even number, but what about the first 10 even numbers after it?
```
> take 10 (filter even [43..])
[44,46,48,50,52,54,56,58,60,62]
```

What is the sum of those?
```
> sum it
530
```
**NOTE**: We used a special feature of GHCi here, which is a special `it` variable that remembers the result of the last expression.

Great, you got the first taste of Haskell! Now let's get to running a real program.

## Writing your first Haskell program

In your editor, create a new file named `hello.hs`.

Write the following in it:
```hs
main = do
  putStrLn "Hello, everybody!"

>   putStrLn ("Please look at my favorite odd numbers: " ++ filter odd [10..20])

Reviewer 2: Wait, this doesn't like like it will compile, `filter odd [10..20]` isn't of `String` type. perhaps add a `let` and `show`?

Author: Ouf, fixed. thanks!

4. **Stack** -> Haskell build tool -> alternative to `cabal`

<div class="bs-callout bs-callout-info">
  <p>
    <h4>cabal and Stack -> which one should I install?</h4>
    cabal is the original build tool, while Stack was created as an alternative to cabal a time ago to solve some of the problems that cabal had and provide a more user-friendly experience. In the meantime, cabal solved most of those issues (including "cabal hell") and caught up with Stack regarding user experience, so the main difference between them at the moment is how they resolve dependencies, which for you as a beginner isn't a concern. Therefore, both are a good choice, and if not sure, you can install both and then use whatever the learning resources you will later use will point you to.
  </p>
</div>

To confirm it is all installed correctly, you can run the following commands:
```
> ghc --version
> haskell-language-server-wrapper --version
> cabal --version                               # If you installed cabal
> stack --version                               # If you installed Stack

> ```

Author: I also removed this, I believe GHCup already gives enough feedback upon successful installation of tools.

### Editor

> HLS (Haskell Language Server), which you just installed via GHCup, is the one that brings all the cool IDE features to editors, so as long as your editor has a decent Haskell language extension that utilizes HLS, you are good.

Author: I also removed this, for the sake of brevity.

It is simple to use, works the same way on all platforms (Linux, macOS, Windows), and gives you one central place/method to take care of your Haskell development setup.

Follow instructions at [GHCup webpage](https://www.haskell.org/ghcup/#) to install GHCup. Then, use it to install the Haskell Toolchain, which consists of:

1. **GHC** -> Haskell compiler. We will use it below to run our examples, but in practice, you will mostly be using a build tool like `cabal` or `Stack` to build your code, instead of `GHC` directly.
2. **HLS** -> Haskell Language Server -> You won't use this directly, instead your code editor will use it in the background to provide you with a great experience while editing Haskell code.
3. **cabal** -> Haskell build tool -> You will use this to structure your Haskell projects, build them, run them, define dependencies, ...
4. **Stack** -> Haskell build tool -> alternative to `cabal`

<div class="bs-callout bs-callout-info">
  <p>
    <h4>cabal and Stack -> which one should I install?</h4>

>     cabal is the original build tool, while Stack was created as an alternative to cabal a time ago to solve some of the problems that cabal had and provide a more user-friendly experience. In the meantime, cabal solved most of those issues (including "cabal hell") and caught up with Stack regarding user experience, so the main difference between them at the moment is how they resolve dependencies, which for you as a beginner isn't a concern. Therefore, both are a good choice, and if not sure, you can install both and then use whatever the learning resources you will later use will point you to.

Reviewer 2: I'm not sure this is the relevant information users look for, I think the answer is "We recommend Installing Both. Most Haskell projects can be built using Cabal, but some might require Stack. Installing both guarantees that you can use either, and follow your tutorial of choice to learn more about them."

Author: OK, went with your suggestion, and skipped mentioning "cabal hell"! Maybe just best not to mention it, similar like you suggested not mention the PhD thing -> that way we don't scare them potentially.

  </p>
</div>

To confirm it is all installed correctly, you can run the following commands:
```
> ghc --version
> haskell-language-server-wrapper --version
> cabal --version                               # If you installed cabal
> stack --version                               # If you installed Stack
```

### Editor
HLS (Haskell Language Server), which you just installed via GHCup, is the one that brings all the cool IDE features to editors, so as long as your editor has a decent Haskell language extension that utilizes HLS, you are good.

> The most straightforward option is to go with **Visual Studio Code** - just install [Haskell extension](https://marketplace.visualstudio.com/items?itemName=haskell.haskell) and you are all set, it should work out of the box and use your installation of HLS.

Reviewer 2: ```suggestion
**Visual Studio Code** is a choice editor with well supported editor integration. Install the [Haskell extension](https://marketplace.visualstudio.com/items?itemName=haskell.haskell) and you are all set, it should work out of the box and use your installation of HLS.
```

> Of other editors that have good Haskell extensions, the most popular ones are Vim and Emacs.

Reviewer 2: I think this is redundant unless the we link to the editor integration plugins/methods - I think users of vim and emacs like to tweak with their editor and will likely try to find information on haskell editor integration with their editor of choice on their own.

Reviewer 2: This is the getting started page, either you help me get started or you're just teasing :)

Author: I don't understand why wouldn't I be helping / how am I teasing if I mention vim and emacs? VSCode, vim and emacs are three most popular editors for Haskell, with good support all of them.

We highlight VSCode because it is the easiest to get started with, it is very popular, and setup is super easy. People who have VSCode already or don't use vim/emacs will follow this happily.

We mention vim and emacs shortly. If you are vim or emacs user, you will be happy to see this, and will investigate further. If you are not vim or emacs user, you will ignore this.

If needed I can remove this line, this is not something I care that much about to compromise the PR for it, but it feels a bit biased to me to mention only vscode.

Reviewer 2: > I don't understand why wouldn't I be helping / how am I teasing if I mention vim and emacs?

What I'm trying to say is that emacs users are going to look for Haskell integration with emacs regardless to whether we list it here or not, because as an emacs user I do all of my programming in emacs regardless of the language. So I find it redundant, *unless* we help them get started. One way to help is to point to the relevant headings in the [hls configuration docs](https://haskell-language-server.readthedocs.io/en/latest/configuration.html).

What I'm trying to say is that the getting started page should not send anyone to google anything. If it's mentioned here, it should include enough context, or include the relevant links to more information. So either we don't write "emacs integration is available" and count on emacs users to google anyway (because they likely will), or we save them the search and send them to the relevant links.

Author: Sounds good @soupi , what I did now is removed the paragraph that mentions emacs and vim and replaced it with
```
To learn about support for other editors, check out [HLS docs for editor configuration](https://haskell-language-server.readthedocs.io/en/latest/configuration.html#configuring-your-editor).
```

Finally, if that still doesn't feel right, I can also remove that, but I think it now follows your recommendation from above.

> ## Running first lines of code

Reviewer 2: I think this section could instead be a link to an existing tutorial, or be reduced to "let's check that we can compile a program with the tools we installed, for me information about the core tools see this tutorial".

We have everything set up, let's use it!

`GHC` brings an interactive interpreter called `GHCi` together with it, which is great for playing with Haskell and trying things out, so let's give it a spin.

> So even though Haskell is a compiled language, it comes with an interpreter too -> how cool is that?

Reviewer 2: compiled is not a property of a programming language but rather of an implementation, so it's not really correct to call Haskell a compiled language. Instead you could say that GHC can compile a Haskell program using the `ghc` command, or interpret it using the `runghc` command. GHC also provides an REPL-style interactive interpreter via the `ghci` command.

Author: Removed

Run `ghci`, which should start a new prompt for you.

Let's do a simple calculation to check Haskell's computing capabilities:
```
> 6 + 3^2 * 4
42
```

> Hey, they call Haskell lazy, but that was quick!

Reviewer 2: "Was it quick? It's just a calculator! And why do "they" call Haskell lazy?" It's fine to leave it out.

Author: Oh well, that was just a joke! I sure have told bad jokes in my life, but this is the first time somebody told me it would be better to not tell it again :D.

I think this relates to the discussion about the "cutesy" tone -> I am ok with removing this joke of course, but if that is ok I would wait just a bit to hear some other opinions.

42 is a nice even number, but what about the even numbers after it? Let's get the first 10 even numbers after 42.
```

> > take 10 $$ filter even [43..]

Reviewer 2: ```suggestion
> take 10 (filter even [43..])
```

Author: I don't mind changing this if that makes sense, but would you mind providing an explanation?

Reviewer 2: https://www.haskellforall.com/2015/09/how-to-make-your-haskell-code-more.html

[44,46,48,50,52,54,56,58,60,62]
```

What is the sum of those?
```
> sum it
530
```
**NOTE**: We used a special feature of GHCi here, which is a special `it` variable that remembers the result of the last expression.

Great, you got the first taste of Haskell! Now let's get to running a real program.

## Writing your first Haskell program

> In your editor, create a new file named `hievb.hs`.

Reviewer 2: ```suggestion
In your editor, create a new file named `hello.hs`.
```

Author: Again, same as above, I don't mind making a change if it makes sense, but I can't make any kind of judgement without an argument being provided.

Reviewer 2: The word hello exists, is descriptive and has meaning. hievb is not a word and I couldn't tell what it means without thinking about it a little, and then saying to myself "why not just hello?".

Write the following in it:
```
main = do
  putStrLn "Hi, everybody!"

>   putStrLn $$ "Pls look at my favorite odd numbers: " ++ filter odd [10..20]

Reviewer 2: ```suggestion
  putStrLn ("Please look at my favorite odd numbers: " ++ filter odd [10..20])
```

```

You can now compile it with `ghc`, which will produce `hievb` binary that we will then run:
```
> ghc hievb.hs
> ./hievb
Hi, everybody!
Pls look at my favorite odd numbers: [11,13,15,17,19]
```

There you go, you just wrote a short, polite program in Haskell!

**TIP**: To interpret the source file directly, without producing any build artifacts, you can use the special `runghc` command like this:
```
> runghc hievb.hs
Hi, everybody!
Pls look at my favorite odd numbers: [11,13,15,17,19]
```

**GHCI TIP**: You can also load your file directly into `ghci`, which will enable you to play with any functions and other definitions you defined in it. So for our example, we can just load hievb.hs with ghci and then call the function `main` like this:
```
> ghci hievb.hs

> ... Compiling ...

Reviewer 2: ?

Author: Hm, I wanted to write something to represent that output that ghci gives here, but didn't want to copy it all so I put this instead. It is a bit weird I agree. What is your suggestion, would you put the whole output?

Btw the output that ghci gives, that I replaced with "... Compiling ...", is:
```
GHCi, version 8.10.7: https://www.haskell.org/ghc/  :? for help
[1 of 1] Compiling Main             ( hievb.hs, interpreted )
Ok, one module loaded.
```

Reviewer 2: Yes, it's better to show to real output of GHCi.

Author: Ok, I put real output of GHCi.

> main

> Hi evb!

Reviewer 2: The output is different from the code I wrote, why?

Author: Well either there is a very specific bug in GHCi, or I made a mistake while writing this PR :D! I bet it is the second. Fixed!

```

> **PRO TIP**: Once you progress a bit more in your Haskell journey, you will very likely want to use `cabal` or `Stack` for building your projects, and not `ghc` directly.

Reviewer 2: Unless you are showing how, it is better to remove this. Their Haskell tutorial of choice should explain that.

Author: Why?

Reviewer 2: Because imo it's not fun to be told "I'm showing you how to do this, but actually you want to do it in a different way" without actually explaining what that way is. I'm just left scratching my head.

Author: Ok, removed!

## Join the community

By joining the Haskell community you will find a great place to ask for help and learn about new developments in the Haskell ecosystem.

Some of the most popular communities are:

 - [r/haskell](https://www.reddit.com/r/haskell/)
 - [Haskell Discourse](https://discourse.haskell.org/)

>  - [https://wiki.haskell.org/IRC_channel](https://wiki.haskell.org/IRC_channel)

Reviewer 2: A short description of each might be in order? I would even replace them with a link to the [Haskell Planetarium](https://haskell.pl-a.net/).

Author: Haskell Planetarium is cool, but it has so many links on it, too many for getting started I think. ALso, is it an official resource? What is wrong with us listing the 3 most popular places here where Haskellers are?

I didn't want to get into descriptions because I felt it is hard to write something without putting one option above the other and therefore getting subjective. Do you have any suggestions on what those descriptions could be?

Reviewer 3: Having a few links seems like a good idea. Too many options would just be confusing for people.

 
We recommend joining right now, and don't be shy to ask for help!

Check [https://www.haskell.org/community](https://www.haskell.org/community) for a full list of Haskell communities.

## Next steps

> There is a myth going around that you need Ph.D. in maths to be able to learn Haskell - and while it is true that Haskell in its elegance does often come close to the elegance of mathematics and borrows a lot of concepts from it, you certainly don't need any advanced math knowledge to be proficient with Haskell. Instead, what you need are some good learning resources!

Reviewer 2: IMO remove this. This is planting the idea in the heads of people even if you're trying to deny it. People that reached here aren't the ones need convincing anyway.

Author: I am ok with removing this if others think the same, but it is a reality that very often Haskell is being perceived this way, I don't think we can do any more harm here, quite the opposite, haskell.org saying that you can do haskell without advanced maths could be a big encorangment to beginners.
Why do you think they don't need convincing? I think a bit convincing can't hurt!

But, as I said, I can certainly easily remove this if others also think it does more harm than good.

Reviewer 2: I expect anyone reading the "get started" guide to be interested in getting started with Haskell, meaning they are have already been convinced to try Haskell out! We should get out of their way and remove all potential obstacles, including them possibly thinking "wait, there are people that think you need a phd to learn Haskell? Maybe this isn't a great idea after all".
